src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp

```diff
@@ -532,6 +621,160 @@ bool ReplicatedMergeTreePartCheckThread::onPartIsLostForever(const String & part
+                return nullptr;
+            }
+
+            LOG_TRACE(log, "Background part check: active parts {}", std::distance(active_parts.begin(), active_parts.end()));
+
+            if (checked_part == MergeTreePartInfo{})
+            {
+                const size_t active_parts_size = std::distance(active_parts.begin(), active_parts.end());
+                std::uniform_int_distribution<size_t> uniform_distribution(0, active_parts_size - 1);
+                const size_t i = uniform_distribution(rndgen);
+                chassert(i < active_parts_size);
+
+                part = *(active_parts.advance_begin(i).begin());
+
+                LOG_TRACE(log, "Background part check: first part to check {}", part->name);
+            }
+            else
+            {
+                auto next_it = storage.data_parts_by_state_and_info.upper_bound(
+                    MergeTreeData::DataPartStateAndInfo{MergeTreeDataPartState::Active, checked_part});
+                if (next_it == storage.data_parts_by_state_and_info.end())
+                    next_it = storage.data_parts_by_state_and_info.begin();
+
+                part = *next_it;
+            }
+
+            MergeTreeDataPartPtr covering_part = storage.getActiveContainingPart(part->info, MergeTreeDataPartState::Active, parts_lock);
+            if (covering_part != part)
+                part = covering_part;
+        }
+
+        if (part->last_check_time.has_value())
+        {
+            const auto check_delay_for_part_seconds
+                = std::chrono::seconds{storage.getSettings()->delay_between_background_part_checks_for_individual_part_seconds};
+            if (part->last_check_time.value() + check_delay_for_part_seconds > std::chrono::steady_clock::now())
+            {
+                checked_part = part->info;
+                part = nullptr;
+            }
+        }
+    }
+
+    return part;
+}
+
+std::chrono::milliseconds ReplicatedMergeTreePartCheckThread::updateBackgroundCheckBackoffTimeout()
+{
+    constexpr auto MIN_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS = std::chrono::milliseconds{100};
+    constexpr auto MAX_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS = std::chrono::milliseconds{10 * 60 * 1000}; // 10 minutes
+
+    auto backoff_timeout_ms = background_check_backoff_timeout_ms;
+    if (backoff_timeout_ms.count())
+    {
+        backoff_timeout_ms *= 2;
+        if (backoff_timeout_ms > MAX_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS)
+            backoff_timeout_ms = MAX_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS;
+    }
+    else
+        backoff_timeout_ms = MIN_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS;
+
+    background_check_backoff_timeout_ms = backoff_timeout_ms;
+    return background_check_backoff_timeout_ms;
+}
+
+std::chrono::milliseconds ReplicatedMergeTreePartCheckThread::delayBetweenChecks()
+{
+    /// backoff timeout is used to avoid overscheduling in case there is not parts to check and background_part_check_delay_seconds is small
```

Comment:
```suggestion
    /// backoff timeout is used to avoid overscheduling in case there is no parts to check and background_part_check_delay_seconds is small
```